### DISCUSSION TITLE
Color Video

### DISCUSSION THREAD

**Discussion by joshuatbland** [LaboratoireMecaniqueLille/crappy#123] [Q&A] [answered]
Hello everyone. I have a script that is collecting video from a webcam and collecting voltage values from a DAQ. For this specific project it would be helpful to collect the video in full color. I made a class based on the 'Webcam' class except it does **not** convert  the image color space to gray scale in the 'get_image' function. Despite this I am still getting saved images that are in grayscale. Any advice or tips on this would be greatly appreciated. Thanks.

**Comment by WeisLeDocto** [comment]
Hi !

Can you please share the script you're running ? And also the custom Camera class that you wrote ?
Also, which version of Crappy are you using ? On which OS ?

**Reply by joshuatbland** [reply]
I just realized that the 'channels' value can be set to 3 which allows the displayer window to show in full color for the native 'Webcam' class. The colors of the images being saved are still off though (although not black and white now). Could it have something to do with OpenCV using BGR as its standard color order and Matplotlib using RGB? I'm using Crappy version 2.0.5 on Windows. This is the block I am using for the camera:
`  cam = crappy.blocks.Camera(
      'Webcam',  
      config=True,  
      display_images=True,  
      displayer_framerate=30,  
      save_images=True,
      img_extension = 'tiff',
      save_folder = data_save_path+'/'+'images',
      save_period = 10,
      freq=40,  
      )`

**Reply by WeisLeDocto** [reply, marked as answer]
I played a bit with my own webcam, and I was able to reproduce your bug.

Basically, several backends are supported in Crappy to save images: SimpleITK, Pillow, OpenCV, and Numpy. This is because depending on the target image format, their performance can vary significantly. Also, back when Crappy was created (about 10 years ago already !), installing Python packages was not as easy as now so having a fallback backend was sometimes necessary.

In your script you do not specify a backend, so by default Crappy will check for the availability of `SimpleITK`, `PIL`, and `cv2` modules in this order, and pick the first one available. `numpy` is always available in case none of the three others are as it's a dependency of Crappy, but can only record `.npy` files.

In addition, I just realized that there are bugs when recording color images, both with the SimpleITK and Pillow backends. With SimpleITK, the shape of the image is wrong (640x480x3 -> 3x640x480). With Pillow, the R and B channels are inverted.

So, my guess regarding your bug is that SimpleITK isn't installed on your machine and Pillow is, therefore Pillow is chosen as the backend for recording, and the red and blue channels get swapped on your images. The most straightforward fix I can think of is simply:
```python
cam = crappy.blocks.Camera('Webcam', config=True, display_images=True, displayer_framerate=30, save_images=True, img_extension='tiff', save_folder=data_save_path+'/'+'images', save_period=10, freq=40, save_backend='cv2')
```

Please tell me if this fix works. We're most of the time using grey level images, for which all backends work fine, therefore this bug never came up. I remember that the OpenCV backend can be quite slow in some cases, double-check that you're able to reach the desired framerate when recording.

I'll fix the identified bugs in the next release, of course.

**Reply by joshuatbland** [reply]
Ok, thanks for the help - this solution worked for me! The framerate is a little slower then what the block specifies, but for our specific use case this is not an issue.

**Reply by WeisLeDocto** [reply]
Good to hear ! 
As soon as version `2.0.6` is released, you'll be able to run your script using other backends as well. This will likely improve the achieved framerate (on my Linux machine `SimpleITK` > `PIL`  > `cv2` for the `.tiff` format). 
This release is not planned before the end of the summer break though.
